File: Algorithms(Python)/Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        d={}
        n=len(s)
        start=0
        result=0
        for i in range(n):
            if s[i] in d:
                start=max(start,d[s[i]]+1)
            result=max(result,i-start+1)
            d[s[i]]=i
        return result

File: Algorithms(Java)/Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n=s.length();
        char[] c=s.toCharArray();
        int start=0;
        int result=0;
        HashMap<Character,Integer> h = new HashMap<Character,Integer>();
        for (int i=0;i<n;i++){
            if (h.containsKey(c[i])){
                start=Math.max(start,h.get(c[i])+1);   //abba
            }
            result=Math.max(result,i-start+1);
            h.put(c[i],i);
        }
        return result;
    }
}
